rust/src/dataset/scanner.rs

```diff
@@ -210,141 +210,187 @@ impl Scanner {
+            let columns_in_filter = column_names_in_expr(predicates.as_ref());
+            let filter_schema = Arc::new(self.dataset.schema().project(&columns_in_filter)?);
+            let remaining_schema = filter_schema.exclude(plan.schema().as_ref())?;
+            if !remaining_schema.fields.is_empty() {
+                // Not all columns for filter are ready, so we need to take them first
+                plan = self.take(plan, &remaining_schema)?;
+            }
+            plan = Arc::new(FilterExec::try_new(predicates.clone(), plan)?);
+        }
+
+        // Stage 3: limit / offset
         if (self.limit.unwrap_or(0) > 0) || self.offset.is_some() {
             plan = self.limit_node(plan);
         }
 
-        let project_schema = Schema::try_from(self.schema()?.as_ref())?;
-        plan = Arc::new(ProjectionExec::try_new(plan, project_schema.into())?);
+        // Stage 4: take remaining columns / projection
+        let output_schema = self.output_schema()?;
+        let remaining_schema = output_schema.exclude(plan.schema().as_ref())?;
+        if !remaining_schema.fields.is_empty() {
+            plan = self.take(plan, &remaining_schema)?;
+        }
+        plan = Arc::new(ProjectionExec::try_new(plan, output_schema.clone())?);
 
-        let session_config = SessionConfig::new();
-        let runtime_config = RuntimeConfig::new();
-        let runtime_env = Arc::new(RuntimeEnv::new(runtime_config)?);
-        let session_state = SessionState::with_config_rt(session_config, runtime_env);
-        Ok(RecordBatchStream::new(
-            plan.execute(0, session_state.task_ctx())?,
-        ))
+        Ok(plan)
     }
 
-    fn filter_knn(
-        &self,
-        knn_node: Arc<dyn ExecutionPlan>,
-        filter_expression: Arc<dyn PhysicalExpr>,
-    ) -> Result<Arc<dyn ExecutionPlan>> {
-        let columns_in_filter = column_names_in_expr(filter_expression.as_ref());
-        let columns_refs = columns_in_filter
-            .iter()
-            .map(|c| c.as_str())
-            .collect::<Vec<_>>();
-        let filter_projection = self.dataset.schema().project(&columns_refs)?;
+    //
+    async fn knn(&self) -> Result<Arc<dyn ExecutionPlan>> {
+        let Some(q) = self.nearest.as_ref() else {
+            return Err(Error::IO("No nearest query".to_string()));
+        };
 
-        let take_node = Arc::new(TakeExec::try_new(
-            self.dataset.clone(),
-            knn_node,
-            Arc::new(filter_projection),
-        )?);
-        self.filter_node(filter_expression, take_node)
+        let column_id = self.dataset.schema().field_id(q.column.as_str())?;
+        let use_index = self.nearest.as_ref().map(|q| q.use_index).unwrap_or(false);
+        let indices = if use_index {
+            self.dataset.load_indices().await?
+        } else {
+            vec![]
+        };
+        let knn_idx = indices.iter().find(|i| i.fields.contains(&column_id));
+        if let Some(index) = knn_idx {
```

Comment:
is this just copy paste? any changes?

Reply:
Moved the filters out, so the filters consolidates with the rest of exec path.